### ISSUE TITLE
LocalSearchDecider should not select NoChangeMove, even with ListVariable(allowsUnassigned=true)

### ISSUE THREAD

**Issue by ge0ffrey** [TimefoldAI/timefold-solver#1242]
(Credit to @TomCools for finding this one)

**Side 1 of the story**

In VRP, with `@PlanningListVariable(allowsUnassigned=true)`, a Construction Heuristic step can choose not to assign a visit. This is represented as a NoChangeMove, because no change happens for that CH step. That is fine.

Local Search's ChangeMoveSelector on the other hand, should never select a NoChangeMove: either it assigns it, unassigns it or changes it. It cannot do nothing for a step.

**Side 2 of the story**

In VRP, with `@PlanningListVariable(allowsUnassigned=true)`, if the CH leaves all visits unassigned (for example because all visits assignments break a hard constraint), the Local Search shouldn't just end with "no doable move" (and a bunch of NoChangeMove bail outs in trace logging), see below.

Why doesn't a move that breaks a hard constraint wins the LS step?

```
14:17:13.567 [main        ] TRACE Bailing out of neverEnding selector (Filtering(FromSolutionPropertyValueSelector(visits))) to avoid infinite loop.
14:17:13.567 [main        ] TRACE         Move (ai.timefold.solver.core.impl.heuristic.move.NoChangeMove@4a3631f8) filtered out by a selection filter (Doable moves only).
14:17:13.567 [main        ] TRACE         Move (ai.timefold.solver.core.impl.heuristic.move.NoChangeMove@4a3631f8) filtered out by a selection filter (Doable moves only).
14:17:13.567 [main        ] TRACE Bailing out of neverEnding selector (Filtering(FromSolutionPropertyValueSelector(visits))) to avoid infinite loop.
14:17:13.567 [main        ] TRACE         Move (ai.timefold.solver.core.impl.heuristic.move.NoChangeMove@4a3631f8) filtered out by a selection filter (Doable moves only).
14:17:13.567 [main        ] TRACE Bailing out of neverEnding selector (Filtering(Union([ListChangeMoveSelector(FromSolutionPropertyValueSelector(visits), ElementDestinationSelector(FromSolutionEntitySelector(Traveler), Filtering(FromSolutionPropertyValueSelector(visits)))), ListSwapMoveSelector(FromSolutionPropertyValueSelector(visits), FromSolutionPropertyValueSelector(visits)), ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.kopt.KOptListMoveSelector@78a773fd]))) after (193060) attempts to avoid infinite loop.
14:17:13.568 [main        ] WARN      No doable selected move at step index (0), time spent (778). Terminating phase early.
14:17:13.568 [main        ] INFO  Local Search phase (1) ended: time spent (778), best score (0hard/0medium/0soft), move evaluation speed (0/sec), step total (0).
14:17:13.570 [main        ] INFO  Solving ended: time spent (778), best score (0hard/0medium/0soft), move evaluation speed (254/sec), phase total (2), environment mode (REPRODUCIBLE), move thread count (NONE).
0hard/0medium/0soft
[]
```

**Comment by TomCools** [comment]
@triceo: Reproducer here: https://github.com/TomCools/nomove-bug-reproducer

**Comment by triceo** [comment]
There are three components to this issue:

1. There is a off-by-one error inside of of the selectors, causing the situation where no doable move was ever found. #1249 fixes that.
2. Having fixed that, another issue was uncovered in our Late Acceptance implementation. Since the CH already produces a better score than local search ever could (0/0/0), the Late Acceptance list is filled with this score. Local Search can not produce a move to beat that, end therefore LA is stuck in an infinite loop, never ending the first step. This issue requires some deeper thinking on how to fix it, and will be fixed in a different PR.
3. The `NoChangeMove`s are necessary in the Move Selector framework to support random nearby selection. They can not be removed. They also do not present an issue - the issue was the off-by-one error, and the NCMs were a red herring, drawing attention away. With Move Streams, this will all change and NCMs will go away.

I will close this issue once a fix for the off-by-one error is merged, and we'll discuss approaches to fixing the issue with LA.
